cpp/include/cugraph/utilities/dataframe_buffer.cuh

```diff
@@ -122,100 +81,73 @@ auto allocate_dataframe_buffer(size_t buffer_size, rmm::cuda_stream_view stream_
-void resize_dataframe_buffer(BufferType& buffer,
+template <typename Type>
+void resize_dataframe_buffer(Type& buffer,
                              size_t new_buffer_size,
                              rmm::cuda_stream_view stream_view)
 {
-  size_t constexpr tuple_size = thrust::tuple_size<T>::value;
-  detail::
-    resize_dataframe_buffer_tuple_iterator_element_impl<T, BufferType, size_t{0}, tuple_size>()
-      .run(buffer, new_buffer_size, stream_view);
-}
-
-template <typename T,
-          typename BufferType,
-          typename std::enable_if_t<std::is_arithmetic<T>::value>* = nullptr>
-void shrink_to_fit_dataframe_buffer(BufferType& buffer, rmm::cuda_stream_view stream_view)
-{
-  buffer.shrink_to_fit(stream_view);
-}
-
-template <typename T,
-          typename BufferType,
-          typename std::enable_if_t<is_thrust_tuple_of_arithmetic<T>::value>* = nullptr>
-void shrink_to_fit_dataframe_buffer(BufferType& buffer, rmm::cuda_stream_view stream_view)
-{
-  size_t constexpr tuple_size = thrust::tuple_size<T>::value;
-  detail::shrink_to_fit_dataframe_buffer_tuple_iterator_element_impl<T,
-                                                                     BufferType,
-                                                                     size_t{0},
-                                                                     tuple_size>()
-    .run(buffer, stream_view);
+  if constexpr (is_std_tuple_of_arithmetic_vectors<Type>::value) {
+    std::apply([new_buffer_size,
+                stream_view](auto&&... args) { (args.resize(new_buffer_size, stream_view), ...); },
+               buffer);
+  } else if constexpr (is_arithmetic_vector<Type, rmm::device_uvector>::value) {
+    buffer.resize(new_buffer_size, stream_view);
+  }
 }
 
-template <typename T,
-          typename BufferType,
-          typename std::enable_if_t<std::is_arithmetic<T>::value>* = nullptr>
-size_t size_dataframe_buffer(BufferType& buffer)
+template <typename Type>
+void shrink_to_fit_dataframe_buffer(Type& buffer, rmm::cuda_stream_view stream_view)
 {
-  return buffer.size();
+  if constexpr (is_std_tuple_of_arithmetic_vectors<Type>::value) {
+    std::apply([stream_view](auto&&... args) { (args.shrink_to_fit(stream_view), ...); }, buffer);
+  } else if constexpr (is_arithmetic_vector<Type, rmm::device_uvector>::value) {
+    buffer.shrink_to_fit(stream_view);
+  }
 }
 
-template <typename T,
-          typename BufferType,
-          typename std::enable_if_t<is_thrust_tuple_of_arithmetic<T>::value>* = nullptr>
-size_t size_dataframe_buffer(BufferType& buffer)
+template <typename Type>
+size_t size_dataframe_buffer(Type& buffer)
 {
-  return std::get<0>(buffer).size();
+  if constexpr (is_std_tuple_of_arithmetic_vectors<Type>::value) {
+    return std::get<0>(buffer).size();
+  } else if constexpr (is_arithmetic_vector<Type, rmm::device_uvector>::value) {
+    return buffer.size();
+  }
```

Comment:
Didn't you see a warning about missing return statement?

See
https://github.com/rapidsai/cugraph/blob/branch-21.10/cpp/include/cugraph/prims/update_frontier_v_push_if_out_nbr.cuh#L80

Here, I am forced to use SFINAE instead of the simpler ```if constepxr``` approach to silence the warning due to an nvcc bug.

Reply:
It would have been a problem but I am returning size_t{} in the next line to avoid that warning.

Reply:
Ah, yeah... here the return types are same so easy to fix!!!